### PR TITLE
[Merged by Bors] - Add support for beaconAPI in `lcli` functions

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

NA

## Proposed Changes

Modifies `lcli skip-slots` and `lcli transition-blocks` allow them to source blocks/states from a beaconAPI and also gives them some more features to assist with benchmarking.

## Additional Info

Breaks the current `lcli skip-slots` and `lcli transition-blocks` APIs by changing some flag names. It should be simple enough to figure out the changes via `--help`.

Currently blocked on #3263.

### CODE REVIEW

File: lcli/src/main.rs [hunk 1]
```diff
@@ -57,59 +56,149 @@ fn main() {
...
-                        .value_name("BEACON_STATE")
+                    Arg::with_name("output-path")
+                        .long("output-path")
+                        .value_name("PATH")
                         .takes_value(true)
-                        .required(true)
+                        .help("Path to output a SSZ file."),
+                )
+                .arg(
+                    Arg::with_name("state-path")
```
Comment (realbigsean) [inline lcli/src/main.rs]: in the `transition-blocks` subcommand this is named `pre-state-path` so might be better to keep it consistent

File: lcli/src/skip_slots.rs [hunk 2]
```diff
@@ -1,58 +1,150 @@
...
+                .map_err(|e| format!("Failed to complete task: {:?}", e))?
+                .ok_or_else(|| format!("Unable to locate state at {:?}", state_id))?
+                .data;
+            let state_root = match state_id {
+                StateId::Root(root) => Some(root),
+                _ => None,
+            };
+            (state, state_root)
+        }
+        _ => return Err("mut supply either --state-file or --beacon-url".into()),
```
Comment (realbigsean) [inline lcli/src/skip_slots.rs]:
```suggestion
        _ => return Err("must supply either --state-path or --beacon-url".into()),
```

File: lcli/src/skip_slots.rs [hunk 3]
```diff
@@ -1,58 +1,150 @@
...
+            info!("State path: {:?}", state_path);
+            let state = load_from_ssz_with(&state_path, spec, BeaconState::from_ssz_bytes)?;
+            (state, None)
+        }
+        (None, Some(beacon_url)) => {
+            let state_id: StateId = parse_required(matches, "state-id")?;
+            let client = BeaconNodeHttpClient::new(beacon_url, Timeouts::set_all(HTTP_TIMEOUT));
+            let state = executor
+                .handle()
+                .ok_or("shut down in progress")?
```
Comment (realbigsean) [inline lcli/src/skip_slots.rs]:
```suggestion
                .ok_or("shutdown in progress")?
```

File: lcli/src/skip_slots.rs [hunk 4]
```diff
@@ -1,58 +1,150 @@
...
+        let mut state = state.clone_with(CloneConfig::committee_caches_only());
+
+        let start = Instant::now();
+
+        if partial {
+            partial_state_advance(&mut state, Some(state_root), target_slot, spec)
+                .map_err(|e| format!("Unable to perform partial advance: {:?}", e))?;
+        } else {
+            complete_state_advance(&mut state, Some(state_root), target_slot, spec)
+                .map_err(|e| format!("Unable to perform partial advance: {:?}", e))?;
```
Comment (realbigsean) [inline lcli/src/skip_slots.rs]:
```suggestion
                .map_err(|e| format!("Unable to perform complete advance: {:?}", e))?;
```

File: lcli/src/transition_blocks.rs [hunk 5]
```diff
@@ -1,125 +1,387 @@
...
+//!
+//! It can load states and blocks from file or pull them from a beaconAPI. Objects pulled from a
+//! beaconAPI can be saved to disk to reduce future calls to that server.
+//!
+//! ## Examples
+//!
+//! ### Run using a block from a beaconAPI
+//!
+//! Download the 0x6c69 block and its pre-state (the state from its parent block) from the
+//! beaconAPI. Advance the pre-state to the slot of the 0x6c69 and apply that block to the
```
Comment (realbigsean) [inline lcli/src/transition_blocks.rs]:
```suggestion
//! beaconAPI. Advance the pre-state to the slot of the 0x6c69 block and apply that block to the
```

File: lcli/src/transition_blocks.rs [hunk 6]
```diff
@@ -1,125 +1,387 @@
...
+            let pre_state = load_from_ssz_with(&pre_state_path, spec, BeaconState::from_ssz_bytes)?;
+            let block = load_from_ssz_with(&block_path, spec, SignedBeaconBlock::from_ssz_bytes)?;
+            (pre_state, None, block)
+        }
+        (None, None, Some(beacon_url)) => {
+            let block_id: BlockId = parse_required(matches, "block-id")?;
+            let client = BeaconNodeHttpClient::new(beacon_url, Timeouts::set_all(HTTP_TIMEOUT));
+            executor
+                .handle()
+                .ok_or("shut down in progress")?
```
Comment (realbigsean) [inline lcli/src/transition_blocks.rs]:
```suggestion
                .ok_or("shutdown in progress")?
```

File: lcli/src/transition_blocks.rs [hunk 7]
```diff
@@ -1,125 +1,387 @@
...
+
+        if output_post_state.is_none() {
+            output_post_state = Some(post_state)
+        }
+    }
 
-    let mut output_file =
-        File::create(output_path).map_err(|e| format!("Unable to create output file: {:?}", e))?;
+    /*
+     * Write artefacts to disk, if required.
```
Comment (realbigsean) [inline lcli/src/transition_blocks.rs]:
```suggestion
     * Write artifacts to disk, if required.
```

**Review by realbigsean — CHANGES_REQUESTED** [review]
Mostly just some typos, these new commands look really useful!

**Comment by paulhauner** [comment]
All comments addressed, I'm going to add "ready for bors" since this is just an lcli command and seems to have implicit approval.

**Comment by paulhauner** [comment]
I'm flagging this as backwards-incompat just so that we remember to mention it in the release notes. I don't think that changes to `lcli` should *actually* be considered "breaking changes" to Lighthouse.

**Pull request closed without merging**